cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll

```diff
@@ -735,6 +737,34 @@ class InitialGlobalValue extends Node, TInitialGlobalValue {
   override string toStringImpl() { result = globalDef.toString() }
 }
 
+/**
+ * A data-flow node used to model flow summaries.
+ */
+class FlowSummaryNode extends Node, TFlowSummaryNode {
+  /**
+   * TODO: QLDoc.
+   */
+  FlowSummaryImpl::Private::SummaryNode getSummaryNode() { this = TFlowSummaryNode(result) }
+
+  /**
+   * TODO: QLDoc.
+   */
+  FlowSummaryImpl::Public::SummarizedCallable getSummarizedCallable() {
+    result = this.getSummaryNode().getSummarizedCallable()
+  }
+
+  /**
+   * TODO: QLDoc.
+   */
+  override Declaration getEnclosingCallable() {
+    result = this.getSummarizedCallable() // TODO: this doesn't look right.
```

Comment:
I think this looks right? The enclosing callable of a flow summary node should be given by the shared flow summary library, and I think you're calling the exact right predicate to grab this information.

Reply:
I think we need to fill out the QLDoc before this will come together.  I'm confused by the presence of both `getSummarizedCallable` and `getEnclosingCallable`, is the former really just supposed to return the enclosing callable as well?

Reply:
I would still like some help writing QLDoc for the `FlowSummaryNode` class.  With QLDoc I should be able to better understand this class and other QL that connects to it (which I'm also a little uncertain of; uncertainty is infectious).
- I can't tell the difference between what the three main predicates are supposed to do.
- I'm not 100% confident of what a `FlowSummaryNode` is anyway.  I think these are all the synthesized nodes that are used to implement a flow summary of a function?  Presumably in practice these form a 'shell' of the function, consisting of inputs and outputs and possibly also nodes representing educated guesses of the implementation (like that a closure passed in to the function is called???)

Reply:
... I am gradually figuring this out.  The `FlowSummaryNode.ql` test and some experimenting helped.
 - the `FlowSummaryNode`s are synthesized nodes used to represent summarized functions, in practice most of them are parameter and return nodes.
 - `getSummarizedCallable` and `getEnclosingCallable` really are the same, with slightly different types.  But they are parts of different interfaces (`getSummarizedCallable` comes from the models-as-data summary node side; `getEnclosingCallable` is just part of `DataFlow::Node`).
 - I've just pushed some QLDoc changes that make this clearer to me at least.

Uncertainties that remain:
 - I think `SummaryPostUpdateNode`s occur when a parameter is modelled as an out param.
 - I think `SummaryOutNode` represents the return value of a synthesized call (`SummaryCall`), when it is needed by the summary.
 - I think `SummaryArgumentNode`s are also only used for synthesized calls (i.e. are uncommon), presumably where the arguments are prepared to go into the `SummaryCall`???  We might be getting a few excess ones in the test actually, for `this` arguments that don't exist.

Reply:
Sorry, I missed this follow-up. Let me answer all of it here. It sounds like there are some very fundamental things that would be good to have an answer to:

`FlowSummaryNode` is the dataflow node that represents all dataflow nodes that are represents internal stuff that happens inside a model. Basically, if a class is a subclass of `FlowSummaryNode` it means it represents something that happens _inside the body of the modelled function_.

What the MaD library does is that it uses the MaD row to synthesize the body of the function as it would look if we had an implementation of it. For example, if we have a MaD row like:
```
;;;foo;;;Argument[0].myField;Argument[1].Parameter[0];summary;
```
this means that the function `foo` does the following:
1. it reads the `myField` field of the 0'th argument, and
2. it passes this value to the 1st argument's 0'th parameter.

In other words, the function `foo` basically must look like:
```cpp
struct MyStruct {
  ...
  int myField;
};
void foo(MyStruct myStruct, void (*f)(int)) {
  f(myStruct.myField);
}
```

That is, the MaD row specifies a kind of canonical function body that can be generated, and all the nodes required to generate this function are represented by `FlowSummaryNode`.

Reply:
Now, let me respond to your latest post:

> `getSummarizedCallable` and `getEnclosingCallable` really are the same, with slightly different types. But they are parts of different interfaces (`getSummarizedCallable` comes from the models-as-data summary node side; `getEnclosingCallable` is just part of DataFlow::Node).

`getSummarizedCallable` is the predicate that gives an enclosing callable to the `FlowSummaryNode`s. Since `FlowSummaryNode` are synthesized from the MaD string, there may very well not be an enclosing callable since the body of the modeled function may not exist in the database. So `getSummarizedCallable` is the way of obtaining the enclosing callable for a `FlowSummaryNode`.


It may that you can answer the other questions you had based on the information I've already given, but let me write it out to be on the safe side.

> I think SummaryPostUpdateNodes occur when a parameter is modelled as an out param.

`SummaryPostUpdateNode` is a `FlowSummaryNode` that is also a `PostUpdateNode`. That is, the MaD row demands the presence of some `PostUpdateNode`, and so the MaD library has synthesized such as node. A `SummaryPostUpdateNode` node is needed when (among other things) the MaD library needs to model a write to a field. For example, if we have a MaD row like:
```
;;;foo;;;Argument[0];Argument[1].myField;summary;
```
which means that the 0'th argument is stored in the 1st argument's `myField` field. Based on this row, MaD will generate `FlowSummaryNode`s to model a call like:
```cpp
struct S {
  int myField;
};
void foo(int x, S* s) {
  s->myField = x;
}
```
and the `storeStep` that is generated from `s->myField = x;` will need to target a `PostUpdateNode`. And because the `storeStep` happens _inside a model_, that store step targets s `SummaryPostUpdateNode`.

> I think SummaryOutNode represents the return value of a synthesized call (SummaryCall), when it is needed by the summary.

Exactly, I hope you can extrapolate my previous answer about `SummaryPostUpdateNode` to this scenario.

> I think SummaryArgumentNodes are also only used for synthesized calls (i.e. are uncommon), presumably where the arguments are prepared to go into the SummaryCall??? We might be getting a few excess ones in the test actually, for this arguments that don't exist.

Correct. `SummaryArgumentNode` are argument nodes that are generated as part of a summary, and they will eventually be used in a `SummaryCall`.

Reply:
I've updated a couple of QLDoc's based on this discussion.

> We might be getting a few excess ones in the test actually, for this arguments that don't exist.

To be more specific, we have a couple of `SummaryArgumentNode`s in the test for `Argument[0].Parameter[this]`, where `Argument[0]` is a function pointer (not a method pointer) so there shouldn't be a `this`.  I suspect these aren't doing any harm but I thought I should mention them.